src/assets/Generator.Shared/ArmOperationHelpers.cs

```diff
@@ -0,0 +1,389 @@
+                if (_info.HeaderFrom == HeaderFrom.Location)
+                {
+                    return response.Status != 202;
+                }
+
+                if (response.Status >= 200 && response.Status <= 204)
+                {
+                    if (response.ContentStream?.Length > 0)
+                    {
+                        try
+                        {
+                            using JsonDocument document = JsonDocument.Parse(response.ContentStream);
+                            foreach (JsonProperty property in document.RootElement.EnumerateObject())
+                            {
+                                if ((_info.HeaderFrom == HeaderFrom.OperationLocation ||
+                                     _info.HeaderFrom == HeaderFrom.AzureAsyncOperation) &&
+                                    property.NameEquals("status"))
+                                {
+                                    state = property.Value.GetString().ToLowerInvariant();
+                                    return s_terminalStates.Contains(state);
+                                }
+
+                                if (_info.HeaderFrom == HeaderFrom.None && property.NameEquals("properties"))
+                                {
+                                    foreach (JsonProperty innerProperty in property.Value.EnumerateObject())
+                                    {
+                                        if (innerProperty.NameEquals("provisioningState"))
+                                        {
+                                            state = innerProperty.Value.GetString().ToLowerInvariant();
+                                            return s_terminalStates.Contains(state);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                        finally
+                        {
+                            // It is required to reset the position of the content after reading as this response may be used for deserialization.
+                            response.ContentStream.Position = 0;
+                        }
+                    }
+
+                    // If provisioningState was not found, it defaults to Succeeded.
+                    if (_info.HeaderFrom == HeaderFrom.None)
+                    {
+                        return true;
+                    }
+                }
+
+                throw _clientDiagnostics.CreateRequestFailedException(response);
+            }
+        }
+
+        private static readonly string[] s_failureStates = { "failed", "canceled" };
+        private static readonly string[] s_terminalStates = s_failureStates.Append("succeeded").ToArray();
+
+        private enum HeaderFrom
+        {
+            None,
+            OperationLocation,
+            AzureAsyncOperation,
+            Location
+        }
+
+        private class ScenarioInfo
```

Comment:
Any reason these can't be fields on the Operation?

Reply:
They are input to create the ArmOperation. We create the ScenarioInfo before making an ArmOperation. I could expose them from the ArmOperation, but we're already in a class, so they are already in the scope where they are used.

Reply:
Sure, but why does this need to be a separate type and why cant' we extract these values in the operation ctor.

Reply:
It was done for conceptual encapsulation. But now, I've changed the code to no longer have this type. Fixed.